### ISSUE TITLE
Search yields "error: column thoughts.quote does not exist"

### ISSUE THREAD

**Issue by sedubois** [thoughtbot/administrate#1837]
* What were you trying to do?

Basically all of our app's content is stored in DB and translated using [Mobility](https://github.com/shioyama/mobility). For instance the `Thought` model has `translates :quote` and `ThoughtDashboard` has `quote: Field::Text`. Listing, showing and editing the thoughts' quotes in the different languages from the admin dashboard works as expected (as in the main app, there is a language switcher so that the values are scoped to the relevant language). This works because behind the scenes, Mobility automatically generates the required getter and setter methods which it maps to ActiveRecord associations.

When using the administrate search box, thoughts which contain a quote containing the provided search term should be returned.

* What did you end up with (logs, or, even better, example apps are great!)?

It yields the error:
```
ActionView::Template::Error: PG::UndefinedColumn: ERROR:  column thoughts.quote does not exist
LINE 1: ...me" AS CHAR(256))) LIKE '%thought%' OR LOWER(CAST("thoughts"...
                                                             ^
```

It appears that because the dashboard is configured with `Field::Text`, Administrate assumes that the data is stored in a text column in the same table. However the data may well be stored in the same table but using JSON, or in a key-value polymorphic table, or (as in our case) in a belongs-to table, or why not even in other ways. Mobility has various backend strategies but provides a unified querying API (i.e. `thought.quote` "just works").

Mobility has an [integration with Ransack](https://github.com/shioyama/mobility-ransack) allowing for case-insensitive partial maches with `Thought.ransack(quote_cont: 'foo').result`. But even without Ransack, it's possible to do a basic exact-match search with `Thought.i18n.where(quote: "blah")`.

* What versions are you running?
  - Rails  6.1.0.rc1
  - administrate master 1a35ab210435995a5e3afcd1efd4163c9aff9f83

**Comment by pablobm** [comment]
This is going to be a tricky one. Administrate's search currently makes some assumptions that don't hold well here. Ideally we would want to create a custom field type (eg: `Field::MobilityI18nText`) that described how to search it. However this is not currently supported, and it's not trivial to add.

An alternative would be creating your own search, instead of using Administrate's default one. Remove Administrate's search field, use your own one, send queries to a custom action, handle it, and render the result with an Administrate `index` view. I think that should work, but I haven't tried.

**Comment by sedubois** [comment]
@pablobm OK below is a custom solution to get ILIKE search working for translated models. 

```diff
# app/controllers/admin/application_controller.rb
module Admin
  class ApplicationController < Administrate::ApplicationController
    #...
    
    def index
      # copied from superclass, with following changes:
+     return super unless resource_class < Translatable
      search_term = params[:search].to_s.strip
-     resources = Administrate::Search.new(scoped_resource,
-                                          dashboard_class,
-                                          search_term).run
+     resources = search_term.blank? ? scoped_resource.all : i18n_search(search_term)
      resources = apply_collection_includes(resources)
      resources = order.apply(resources)
      resources = resources.page(params[:page]).per(records_per_page)
      page = Administrate::Page::Collection.new(dashboard, order: order)

      render locals: {
        resources: resources,
        search_term: search_term,
        page: page,
        show_search_bar: show_search_bar?,
      }
    end

    #...
```
```ruby
    def i18n_search(search_term)
      search_term = sanitize_sql_like(search_term)
      attributes = dashboard_class::ATTRIBUTE_TYPES.select { |_, field| field.searchable? }

      resource_class.i18n do
        def match((attribute, field), search_term)
          # TODO add API to fields to allow matchers for other column types, e.g. `.eq(search_term)`
          instance_eval(attribute.to_s).matches("%#{search_term}%")
        end

        attributes.reduce(nil) do |memo, nxt|
          memo.nil? ? match(nxt, search_term) : memo.or(match(nxt, search_term))
        end
      end
    end
  end
end
```

Actually the method is called `i18n_search()`, but except for the scope `.i18n` coming from Mobility it is quite generic (and I already apply it using a `default_scope`). I just don't know how to replace that scope to perform the Arel query on the class' default scope, so that it would also work in non-translated models. Any idea how to make that generic?

The great thing with the solution above is that thanks to Mobility's API it should work no matter how the translation data is stored (in a key-value table, in a JSON column, etc).

Anyway at least now I have something working for my app. Next steps could be to retrofit something like this in Administrate, and also play with having more proper full-text search with pg_search or searchkick.

**Comment by sedubois** [comment]
@shioyama any idea about this question above?

> Actually the method is called i18n_search(), but except for the scope .i18n coming from Mobility it is quite generic (and I already apply it using a default_scope). I just don't know how to replace that scope to perform the Arel query on the class' default scope, so that it would also work in non-translated models. Any idea how to make that generic?

**Comment by shioyama** [comment]
Hmm I don't know administrate at all, so it's going to be hard to say very much. Under the hood the `i18n` scope in Mobility is doing two things, creating an Arel node for each translated attribute:

```ruby
backend_class = Post.mobility_backend_class(:title)
node = backend_class.build_node(:title, :en)
```

You can do things like `node.eq('foo')` etc like any other Arel node.

Once you have the node, you the build a relation by applying the backend scope to a relation (like `Post.all`):

```ruby
relation = backend_class.apply_scope(Post.all, node, :en)
```

You can then call `relation.all`, `relation.first`, `relation.to_sql` etc.

That's pretty complicated which is why Mobility gives you the ability to just pass a block to the query scope like this:

```ruby
Post.i18n do
  title.eq('foo')
end
```

The `title` in the block resolves to the `node` above.

If you need to do something tricky, you might need to actually build up the relation using the steps above.

**Comment by sedubois** [comment]
Thanks @shioyama, my question wasn't clear. I was wondering how to change `Post.i18n` to something generic, in the case when `Post` already has `default_scope { i18n }`. I thought the answer could be to replace `i18n` with `default_scoped`:

```diff
-     resource_class.i18n do
+     resource_class.default_scoped do
```

However that doesn't work, it skips the `default_scope { i18n }` defined in `Post`. I just don't know how to navigate the ActiveRecord/Mobility codebase to find the answer to this myself.

Basically, in a plain Rails project (where Mobility—or Administrate for that matter—is not involved at all), how to make this query work?

```rb
Post.??? { title.matches("%blah%") }
```

**Comment by shioyama** [comment]
Hmm, so actually you made me realize that some things in the query plugin that are currently private should actually be exposed.

But to answer your question, there is no `???` which will work AFAIK in Rails, since Rails does not support block-format query. Mobility custom implements it using a class called `Mobility::Plugins::ActiveRecord::Query::VirtualRow`, which is actually private but I will make it public in the next release.

You _can_ do this currently but it's going to be quite ugly:

```ruby
Mobility::Plugins::ActiveRecord::Query.const_get(:VirtualRow).build_query(resource_class, Mobility.locale) do
  title.matches("%foo%")
end
```

I think that might do what you want. I will make the `VirtualRow` class public in the next release so you don't need that `const_get` (which gets around `private_constant`).

**Comment by shioyama** [comment]
Ok I created https://github.com/shioyama/mobility/pull/471 which will make it slightly easier to do this:

```ruby
Mobility::Plugins::ActiveRecord::Query.build_query(resource_class) do
  title.matches("%foo%")
end
```

**Comment by sedubois** [comment]
> there is no ??? which will work AFAIK in Rails, since Rails does not support block-format query

Do you think a case could be made to request adding this in Rails? If there is a reason to refuse it, then the flip-side of that question would be, how to make queries in Mobility which match "the Rails way"? Or would you recommend I just give up on this idea of having the same API in both?

**Comment by shioyama** [comment]
> Do you think a case could be made to request adding this in Rails? 

This is basically what [Wharel](https://github.com/shioyama/wharel) does, and I believe there is a PR somewhere to make Rails do it as well (might have been merged actually).

**Comment by sedubois** [comment]
You are possibly referring to [this PR](https://github.com/rails/rails/pull/39445), but it seems far from ready for merge, and anyway would take a long time before being available in all supported Rails versions.

**Comment by 9mm** [comment]
So... is expected behavior that this also breaks a very simple virtual field?

```
    ballistic_coefficient: Field::String,
```

and I have a method on the model (not from DB)
